### ISSUE TITLE
Switch to Golang updater

### ISSUE THREAD

**Issue by SrKomodo** [overdodactyl/ShadowFox#127]
Alright, the time has finally come to unveil the brand new cross-platform [ShadowFox Updater](https://github.com/SrKomodo/shadowfox-updater)!

You can download it [here](https://github.com/SrKomodo/shadowfox-updater/releases/latest)

Before making this the official updater a few changes need to be done to the repo:

1. webextension_mappings.txt has to be removed and the UUIDs need to be changed in all of the css as discussed in #126 (i already did this on https://github.com/SrKomodo/ShadowFox, gonna submit a pull request soon)
2. Once that is done the updater will work, but we will still need to change the UUID bash scripts so that the gulpfile still works (I don't fully understand how the workflow works so i'm leaving that up to someone else to do)
3. After that we just need to do some cleaning up by removing the old updaters, and changing the wiki

Please inform me of any errors or bugs. I only tried it on windows but it should work everywhere

**Comment by overdodactyl** [comment]
The new updater's looking great!! Really excited about this.  I gave it a run on macOS and ran into two little snags, but I've created a pull with fixes

A couple initial thoughts/questions:

* I'm a little hesitant to get rid of `internal_UUIDs.txt` completely as it gets rid of the ability to easily customize which extensions get styled and which don't.  See https://github.com/overdodactyl/ShadowFox/issues/84#issuecomment-372136996 for a little more detail on that.  I know there are some users who prefer not to have all of them styled or just like to have fixes for ones that have compatibility problems.

* Would it be easy enough to create a warning or something similar when a user chooses to uninstall ShadowFox?  Something to the effect of "Warning: this will delete your entire chrome folder and can not be undone."  I'm thinking about users who might have extra tweaks in their chrome directory either in the customization files or `userChrome.js` or something who might not expect them to be deleted as well.  This might be nice since you can't simply open up your Trash folder and retrieve it. I'm not really familiar with Golang yet, so not exactly sure what's easy and what's not.  

* Are you thinking of merging the shadowfox-updater repo with this one or do you want to keep it separate?  I can add you as a collaborator on here so you can easily edit this repo.  It might be nice to have everything together and scripts in the releases section here, but if you want to keep them separate I totally understand!

**Comment by SrKomodo** [comment]
* I implemented support for `internal_UUIDs.txt`. Now the program will get the UUIDs from there, and if the file is empty it will populate it with UUIDs
* I made it so that, when uninstalling, the program will only remove `userContent.css`, `userChrome.css` and `ShadowFox_customization`, this should stop the program from removing files it shouldn't remove
* Because of the way Golang works, it's better if they are kept on separate repos
* Just submitted the PR #128

**Comment by overdodactyl** [comment]
>Now the program will get the UUIDs from there, and if the file is empty it will populate it with UUIDs

Would it be possible to maybe have a drop down like the profile selector to choose between "Auto Generate UUIDs" (selected by default if possible) and "Use internal_UUIDs.txt" or something of that nature?

I'm thinking of the use case where I'm guessing most people want all styling included, so it would be nice if `internal_UUIDs.txt` was created on each update to include any new styling, but this would give the option for a user to manually maintain the file as well.

Everything else sounds great!

**Comment by SrKomodo** [comment]
Alright, now the program has a checkbox for autogenerating the UUIDs, toggled on by default

**Comment by overdodactyl** [comment]
I think the scripts and documentation is just about ready - are you good to go live with these changes?

**Comment by SrKomodo** [comment]
I think everything is ready so yeah, let's do it!

**Comment by overdodactyl** [comment]
@CarlColijn and @Shourai, I do apologize for switching things up a bit, both you guys helped me out a ton and I know put in a lot of time with those other updaters - which I really I do appreciate.  

In the end though I think having a cross-platform solution is a lot more ideal. It was difficult trying to maintain feature parity and consistency without everyone having access to all the platforms.

**Comment by CarlColijn** [comment]
One small change request: the old installer scripts added a timestamp to the backup filenames; this meant I could run the updater multiple times and get as many backups side by side, and it added nice-to-have info to the backup filename.  Now (I assume) the last backup overwrites the previous one?  Should be an easy fix I think?

**Comment by CarlColijn** [comment]
One other small suggestion, which I have no idea of if it's an easy fix: tab switches to the next control, but shift-tab does nothing.  Might that be bound to switching to the prev. control?  Or is that not a standard keyboard shortcut on Linux/Mac?

**Comment by SrKomodo** [comment]
Also, @t2YU2m8l83 I am having a busy day but i hope that tomorrow i can fix that issue

**Comment by overdodactyl** [comment]
@CarlColijn -

>And I also do have some extensions that are not yet themed

Feel free to open up new issues for these and I'll get to them when I have a chance...I owe ya a little for all the help after all :)

And you've made some great suggestions for the project, so don't hesitate to make any more if you have any in the future!

@SrKomodo -

Thanks for all these quick fixes!

**Comment by SrKomodo** [comment]
@t2YU2m8l83 Finally fixed the issue, now in v1.4 if the program can't find profile.ini it will ask you to provide a path

**Comment by t2YU2m8l83** [comment]
@SrKomodo I would prefere if the updater checks for Firefox in the following order:

1. Am I placed inside a profile folder
2. Are profiles under %AppData%
3. Let the user choose a profile path

Also right-click+paste doesn't work for the path and it's pretty long in my case. Ctrl+v should also be an option alongside a file explorer window.

**Comment by SrKomodo** [comment]
The program really needs profiles.ini so that it works, so I suggest you make a profiles.ini file yourself.

Let's imagine you have this folder structure (of course you would adapt this example to your actual folder structure)
```
Firefox
├ Profiles
│ ├ profileA
│ └ profileB
├ profiles.ini
└ shadowfox.exe
```
Then your `profiles.ini` should look like this:
```
[ProfileA]
Path=Profiles/profileA
IsRelative=1

[ProfileB]
Path=Profiles/profileB
IsRelative=1
```
I hope this fixes your problem @t2YU2m8l83 , sorry for the late response

**Comment by t2YU2m8l83** [comment]
@SrKomodo Thanks. I've created "[...]FirefoxPortable\Data\profile\profiles.ini" alongside "[...]\FirefoxPortable\Data\profile\shadowfox_windows_amd64.exe" and wrote the following things into it.
```
[ProfileA]
Path=
IsRelative=1
```
With this changes I can use the isntaller. You should mention this installation fix for FirefoxPortable somewhere.
